python/shotgun_model/shotgun_hierarchy_model.py

```diff
@@ -119,8 +134,141 @@ def item_from_path(self, path):
+        """
+        paths = self._resolve_entity_paths(entity)
+        self.async_load_paths(paths)
+
+    def _resolve_entity_paths(self, entity):
+        """
+        Resolves an entities path in the nav hierarchy.
+
+        .. note::
+            This method is executed in the current thread.
+
+        :param dict entity: Entity dictionary with keys ``type`` and ``id``.
+
+        :returns: The path to the entity in the nav api.
+        :rtype: str
+        """
+        if entity:
+            # FIXME: Unfortunately we can't call the endpoint directly because there is a bug in it.
+            # We've written a workaround for it in the ShotgunDataRetriever, which we will be
+            # using here.
+
+            sg_result = self._sg_data_retriever._task_execute_nav_search_entity("/", entity)["sg_result"]
+
+            if len(sg_result) == 0:
+                logger.warning("Entity %s not found. Picking /.", entity)
+            else:
+                sg_data = sg_result[0]
+                # The last link in the chain is always the complete link to the entity we seek.
+
+                if len(sg_result) > 1:
+                    logger.info(
+                        "Entity %s found %d times with nav_search_entity endpoint. Picking %s.",
+                        entity, len(sg_result), pprint.pformat(sg_data)
+                    )
+                    logger.info("Other choices were %s", sg_result[1:])
+
+                return sg_data["incremental_path"]
+
+        # Fallback for when the root was requested or no entity was found.
+        # Do not request the server for the path to the site root, this will always be /.
+        return ["/"]
+
+    def async_item_from_paths(self, paths):
+        """
+        Takes a list of paths that incrementally dig deeper into the
+        model and signals when the node is found and loaded in memory.
+
+        :param list(str): List of paths from the nav api that drill down further
+            and further into the tree.
+
+        :signals: ``async_item_retrieval_completed``
+        """
+        # Nothing to async load, return early.
+        if not paths:
+            return
+
+        logger.debug("Async loading of %s", paths)
+
+        for idx, path in enumerate(paths):
+            # Iterate on every path.
+            item = self.item_from_path(path)
+            # If an item is already loaded, move to the next one.
+            if item:
+                continue
+            # Send a refresh request and return. When the node is refreshed,
+            # this method will be called a second time with the same parameters.
+            # This time around this node will already have been refreshed
+            # and the code will dig deeper. At some point the last entry
+            # in the list will be reached and we will emit the item.
+            logger.debug("Refreshing paths: %s", paths[idx - 1:])
+            self._NodeRefresher(paths[idx - 1:], self)
```

Comment:
for clarify maybe we could do `self._refreshed= self._NodeRefresher` ? the model child relationship is an internal C++ binding only, always good to keep a python ref too for good measure IMO :)